index.bs

```diff
@@ -1294,44 +1294,101 @@ A {{ReceiveStream}} has the following internal slots.
    <td class="non-normative">An [=incoming unidirectional=] or [=bidirectional=]
    [=WebTransport stream=].
   </tr>
+  <tr>
+   <td><dfn>\[[Transport]]</dfn>
+   <td class="non-normative">The {{WebTransport}} object owning this {{ReceiveStream}}.
+  </tr>
 </table>
 
 ## Procedures ##  {#receive-stream-procedures}
 
-<div algorithm="create a ReceiveStream">
+<div algorithm>
 
 To <dfn export for="ReceiveStream" lt="create|creating">create</dfn> a
 {{ReceiveStream}}, with an [=incoming unidirectional=] or [=bidirectional=] [=WebTransport stream=]
 |internalStream| and a {{WebTransport}} |transport|, run these steps:
 
 1. Let |stream| be a [=new=] {{ReceiveStream}}, with:
-    : [=[[InternalStream]]=]
+    : [=ReceiveStream/[[InternalStream]]=]
     :: |internalStream|
-1. [=ReadableStream/Set up=] |stream| with TBD.
+    : [=ReceiveStream/[[Transport]]=]
+    :: |transport|
+1. Let |pullAlgorithm| be an action that [=pulls bytes=] from |stream|.
+1. Let |cancelAlgorithm| be an action that [=ReceiveStream/cancels=] |stream|.
+1. [=ReadableStream/Set up with byte reading support=] |stream| with
+   [=ReadableStream/set up with byte reading support/pullAlgorithm=] set to |pullAlgorithm| and
+   [=ReadableStream/set up with byte reading support/cancelAlgorithm=] set to |cancelAlgorithm|.
+1. Set |stream|'s [=ReadableByteStream/[[controller]]=]'s
+   [=ReadableStreamController/autoAllocateChunkSize=] to an [=implementation-defined=] integer.
 1. [=set/Add=] |stream| to |transport|'s [=[[ReceiveStreams]]=].
 1. Return |stream|.
 
 </div>
 
-When a {{ReceiveStream}} receives a message from the remote side aborting the
-stream (for QUIC, that message is an RESET_STREAM frame), the user agent MUST
-queue a task to run the following:
-      1. Let |stream| be the {{ReceiveStream}} object for which the reset
-         message was received.
-      1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-         from.
-      1. Let |reason| be the value from the message from the remote side.
-      1. [=ReadableStream/Error=] |stream| with |reason|.
-
-The {{ReceiveStream}}'s <dfn>cancelAlgorithm</dfn> is:
-     1. Let |stream| be the {{ReceiveStream}} object.
-     1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-        from.
-     1. Let |reason| be the first argument, if it is numeric, or 0.
-     1. If |stream| has received a RESET_STREAM frame, abort these steps.
-     1. [=In parallel=], [=send STOP_SENDING=] over |stream|'s
-        [=WebTransport stream=] with |reason| as the Application Protocol Error
-         Code.
+<div algorithm>
+
+To <dfn for="ReceiveStream">pull bytes</dfn> from a {{ReceiveStream}} |stream|, run these steps.
+
+1. Let |controller| be |stream|'s [=ReadableStream/[[controller]]=].
```

Comment:
This algorithm uses too many internals, which should not be necessary. Instead you should use https://whatpr.org/streams/1130.html#readablestream-current-requested-number-of-bytes and https://whatpr.org/streams/1130.html#readablestream-enqueue-bytes as appropriate.

If those don't work for you, then that would be great feedback for the streams spec.

Reply:
Thank you! 

That will lead to additional bytes copy. What do you think? We could say that implementers can always perform an optimization, but...

Reply:
Right, the "implementers can always perform an optimization" is what I am currently proposing in the streams spec. There it says

> Implementations are encouraged to, when possible, do this directly from whatever underlying I/O system is producing _bytes_. In this sense the existence of an intermediate _bytes_ variable might end up as spec fiction, if the underlying I/O system can write directly into the backing data block.

This seems worthwhile so that specs can be written more simply. Note that autoAllocateChunkSize doesn't work for all cases; see some of the discussion in https://github.com/whatwg/streams/issues/1121#issuecomment-814473948 .

I'm very open to feedback though if you think this is too clever and we should be more explicit in the client specs like WebTransport, with less magic in the Streams spec.

Reply:
Thanks, I can live with that.

Another possible approach is, to have "enqueue bytes" take an ArrayBufferView _view_ instead of a byte sequence, and use _byobRequest_ when _view_ is contained in _byobRequest_'s \[[view]].

It will look like:

4. If _byobRequest_ is non-null, _byobRequest_.[[view]].[[ViewedArrayBuffer]] is _view_.[[ViewedArrayBuffer]], 
       _byobRequest_.[[view]].[[ByteOffset]] is _view_.[[ByteOffset]] and 
       _byobRequest_.[[view]].[[ByteLength]] ≤ _view_.[[ByteLength]]:
   1. Perform ! ReadableByteStreamControllerRespond(stream.[[controller]], _view_.[[ByteLength]]). 
5. Otherwise:
   1. Perform ! ReadableByteStreamControllerEnqueue(stream.[[controller]], _view_).

And also expose the byobRequest's \[[view]] as "current requested view" instead of "current requested number of bytes".

I prefer this approach because it matches with my mental model better: when the byobRequest is provided, I can borrow the ArrayBufferView, update the contents and return it or its subarray. Otherwise I can create a new ArrayBufferView, update its contents and enqueue it.

Reply:
That makes sense to me. I will work on updating the spec in that direction.